doc/man/gnunet-nat-auto.1
.\" This file is part of GNUnet.
.\" Copyright (C) 2001-2019 GNUnet e.V.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the file
.\" FDL-1.3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/fdl.html}.
.\"
.\" Alternately, this document is also available under the General
.\" Public License, version 3 or later, as published by the Free Software
.\" Foundation.  A copy of the license is included in the file
.\" GPL3.
.\"
.\" A copy of the license is also available from the Free Software
.\" Foundation Web site at http://www.gnu.org/licenses/gpl.html
.\"
.\" SPDX-License-Identifier: GPL3.0-or-later OR FDL1.3-or-later
.\"
.Dd January 6, 2017
.Dt GNUNET-NAT-AUTO 1
.Os
.Sh NAME
.Nm gnunet-nat-auto
.Nd autoconfigure and test NAT traversal
.Sh SYNOPSIS
.Nm
.Op Fl a | -auto
.Op Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
.Op Fl S Ar NAME | Fl -section= Ns Ar NAME
.Op Fl t | -tcp
.Op Fl u | -udp
.Op Fl w | -write
.Sh DESCRIPTION
.Nm
allows testing various NAT traversal functions, as well as attempting auto-configuration.
The options are as follows:
.Bl -tag -width indent
.It Fl a | -auto
Attempt auto-configuration for NAT traversal.
.It Fl c Ar FILENAME | Fl -config= Ns Ar FILENAME
Use the configuration file
.Ar FILENAME .
.It Fl S Ar NAME | Fl -section= Ns Ar NAME
Name of the configuration section with details about the configuration to test.
For example "transport-tcp".
.It Fl t | -tcp
Use TCP.
.It Fl u | -udp
Use UDP.
.It Fl w | -write
Write configuration to configuration file, useful in combination with auto-configuration
.Pq Fl a .
.El
.Sh EXAMPLES
.Ss Automatic configuration:
.Pp
.Dl # gnunet-nat-auto -aw
.Pp
Probe and write result to configuration
.Ss Test configuration:
.Pp
.Pp
.Dl # gnunet-nat-auto -t -S transport-tcp
.Pp
Test TCP configuration
.Pp
.Dl # gnunet-nat-auto -t -S transport-http
.Pp
Test HTTP configuration
.Pp
.Dl # gnunet-nat-auto -u -S transport-udp
.Pp
Test UDP configuration
.\".Sh FILES
.Sh SEE ALSO
.Xr gnunet-nat 1 ,
.Xr gnunet-transport 1
.sp
The full documentation for gnunet is maintained as a Texinfo manual.
If the
.Xr info 1
and gnunet programs are properly installed at your site, the command
.Pp
.Dl info gnunet
.Pp
should give you access to the complete handbook,
.Pp
.Dl info gnunet-c-tutorial
.Pp
will give you access to a tutorial for developers.
.sp
Depending on your installation, this information is also available in
.Xr gnunet 7 and
.Xr gnunet-c-tutorial 7 .
.\".Sh HISTORY
.\".Sh AUTHORS
.Sh BUGS
Report bugs by using
.Lk https://bugs.gnunet.org
or by sending electronic mail to
.Aq Mt gnunet-developers@gnu.org .


ChangeLog
Wed Oct 30 10:50:00 UTC 2019
  Released gnunet 0.11.8
  -ng0

Sun Oct 27 17:17:00 2019 +0200
  Released gnunet 0.11.7
  -schanzen, ng0

Fri Oct 25 00:00:00 UTC 2019
  Dropped glpk dependency(-check). -ng0

Wed Oct 23 20:55:12 2019 +0200
  Make REST service start per user in a multiuser setup.
  -schanzen

Wed Oct 23 00:00:00 UTC 2019
  Remove setuid helpers. They never worked as intended. The
  fixed version is not portable and the defacto good way to
  handle this which is most portable is to let package managers
  handle this (as they do right now, the good ones), and let
  people handle this who read the documentation.
  This commit removes what would be patched out by the majority
  of package managers, which is setuid handling in Makefiles.
  It is very likely that no one will notice this code is gone.
  -ng0

Thu Oct 17 00:00:00 UTC 2019
  Added support for doas, use it in some places conditionally
  if sudo is not present. Fixed sudo detection. -ng0

Mon Oct 14 00:00:00 UTC 2019
  Added 'pretty' make rule to run uncrustify over the source tree.
  -ng0

Mon Sep 16 00:00:00 UTC 2019
  Remove checks for CXX and CPP. Scans found no
  (remaining) cxx code. ccache with clang on
  NetBSD 9 hit no rebuilds. -ng0

Mon Sep 10 00:00:00 UTC 2019
  Drop win32 and mingw support after discussion with
  Christian. It has been unmaintained in our code base
  for years and there are no known users.
  Future implementors for Windows 10 support can
  refer to this commit as reference. -ng0

Mon Sep 9 00:00:00 UTC 2019
  Remove plibc, closes #5877 as discussed in
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00002.html
  -ng0

Sun Sep 1 13:25:07 2019 +0200
  Improved support for external applications/services wrt
  localization/gettext.
  https://lists.gnu.org/archive/html/gnunet-developers/2019-09/msg00000.html

Tue 20 Aug 2019 04:32:09 PM CEST
  Fixed fatal bug in gnunet-bcd that prevented the main page
  from showing up. -CG

Mon Aug 19 17:36:48 2019 +0200
  REST: Added API plugin to read/write GNUnet configuration options -alexia

Sun Jul 28 00:00:00 2019 +0000
  Adding --enable-texinfo4 configure switch, defaulting to off.
  This allows distributors to override the texinfo 5.x and later
  commandline options we use, and allows us on the serverside to
  build html with newer output options. -ng0

Wed 24 Jul 2019 05:43:43 PM CEST
  Releasing GNUnet 0.11.6. -CG

Mon Jul 15 09:57:20 2019 +0200
  The REST server now has a configurable CORS Origin setting -schanzen

Fri Jul 5 22:11:04 2019 +0200
  Added detection of SUID binaries
  through GNUNET_OS_get_suid_binary_path -lurchi

Sat 29 Jun 2019 10:18:55 AM CEST
  Renamed "heap" namestore backend to "flat" as it does
  persist data in a file. Made it work with databases larger
  than 16 MB. -CG

Fri Jun 28 14:10:53 2019 +0200
  RECLAIM: Reissue of tickets (same attributes, same identity)
  will no longer create a new ticket, but reuse the old. -alexia

Thu Jun 27 10:34:46 2019 +0200
  The REST server now returns the Origin header value in the
  CORS header reponse. -schanzen

Thu Jun 27 09:08:06 2019 +0200
  GNS NSS plugin will not timeout and not hang forever and block
  gethostbyname. Final resolution pending #5782 -schanzen

Sat Jun 22 10:38:43 2019 +0200
  Add gnutls support for GNS CA certificate generation -schanzen

Sat 15 Jun 2019 11:45:35 PM CEST
  Improved gnunet-gns performance for many ego scenario. -CG

Fri 14 Jun 2019 07:17:40 PM CEST
  Add option to gnunet-identity to display private keys. -CG

Wed Jun 5 21:19:46 2019 +0200
 Releasing GNUnet 0.11.5. -schanzen

Wed May 29 11:17:49 2019 +0200
  Added CAA record type (RFC 6844). -schanzen

Sat 25 May 2019 08:27:39 AM CEST
  Significantly speed-up ego creation and deletion via
  gnunet-identity command when there are many (hundreds)
  of egos at the same peer. -CG

Sun May 12 12:00:00 2019 +0200
  Releasing GNUnet 0.11.4. -MS

Thu May 9 23:35:24 2019 +0200
  gnunet-identity can now output the public key on creation of new
  egos, and -e can act as a filter when listing all egos. -CG

Thu May 2 08:44:37 2019 +0200
  gnunet-arm now logs to file by default -CG

Sat Apr 27 21:50:57 2019 +0200
  RECLAIM: Added ECDH encryption for authorization code payload -schanzen

Sat 27 Apr 2019 06:36:31 AM CEST
  Bump libmicrohttpd version requirement as some intermediate
  versions have relevant bugs that cause problems. -CG

Sat Apr 27 01:50:43 2019 +0200
  RECLAIM: No longer experimental -schanzen

Fri Apr 26 11:08:21 2019 +0200
  REST: Configuration plugin added -schanzen

Tue Apr 23 23:00:00 UTC 2019
  All manpages are properly licensed. -ng0

Sun Apr 21 22:22:22 UTC 2019
  All manpages are now in mdoc format. -ng0

Sat Apr 20 18:38:43 UTC 2019
  Remove optional gnunet-download-manager.scm and
  with it the optional dependency on a no longer
  maintained version of GNU Guile (#5682). -ng0

Sun Apr 14 09:54:20 2019 +0200
  RECLAIM: Moved away from ABE to GNS encryption. Minor API changed and
  renaming of records. Deprecated SQLite plugin in favor of namestore.
  -schanzen

Sun Apr 7 14:40:59 2019 +0200
  Releasing GNUnet 0.11.3. -CG

Sat 06 Apr 2019 02:36:17 PM CEST
  Introducing GNUNET_memcmp() and GNUNET_is_zero() for better typing
  when comparing memory areas. -CG

Fri 05 Apr 2019 11:38:53 PM CEST
  Use paths of sysctl, ip, ifconfig and ip(6)tables from configure
  if they work. (#5352). -CG

Thu 04 Apr 2019 12:39:44 PM CEST
  Hotfix for build issue. Releasing GNUnet 0.11.2. -CG/MS

Wed 03 Apr 2019 03:18:43 PM CEST
  Releasing GNUnet 0.11.1. -CG

Wed 03 Apr 2019 01:54:44 PM CEST
  Merged gnunet-qr-c-2 branch, replacing Python-based implementation
  of gnunet-qr with C-based implementation that also integrates
  gnunet-uri. -HG/MS/CG

Wed 03 Apr 2019 01:15:57 PM CEST
  Fixed a few test case regressions (GNS, FS). -CG

Mon Apr 1 22:00:00 UTC 2019
  documentation/buildsystem: manpages and other documentation parts
  can now be build more controlled again, this fixes the regression

src/nat-auto/gnunet-nat-auto.c
/*
     This file is part of GNUnet.
     Copyright (C) 2015, 2016, 2017 GNUnet e.V.

     GNUnet is free software: you can redistribute it and/or modify it
     under the terms of the GNU Affero General Public License as published
     by the Free Software Foundation, either version 3 of the License,
     or (at your option) any later version.

     GNUnet is distributed in the hope that it will be useful, but
     WITHOUT ANY WARRANTY; without even the implied warranty of
     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
     Affero General Public License for more details.

     You should have received a copy of the GNU Affero General Public License
     along with this program.  If not, see <http://www.gnu.org/licenses/>.

     SPDX-License-Identifier: AGPL3.0-or-later
 */

/**
 * @file src/nat/gnunet-nat-auto.c
 * @brief Command-line tool for testing and autoconfiguration of NAT traversal
 * @author Christian Grothoff
 * @author Bruno Cabral
 */
#include "platform.h"
#include "gnunet_util_lib.h"
#include "gnunet_nat_service.h"
#include "gnunet_nat_auto_service.h"

/**
 * Value to return from #main().
 */
static int global_ret;

/**
 * Handle to ongoing autoconfiguration.
 */
static struct GNUNET_NAT_AUTO_AutoHandle *ah;

/**
 * If we do auto-configuration, should we write the result
 * to a file?
 */
static int write_cfg;

/**
 * Configuration filename.
 */
static const char *cfg_file;

/**
 * Original configuration.
 */
static const struct GNUNET_CONFIGURATION_Handle *cfg;

/**
 * Adapter we are supposed to test.
 */
static char *section_name;

/**
 * Should we run autoconfiguration?
 */
static int do_auto;

/**
 * Handle to a NAT test operation.
 */
static struct GNUNET_NAT_AUTO_Test *nt;

/**
 * Flag set to 1 if we use IPPROTO_UDP.
 */
static int use_udp;

/**
 * Flag set to 1 if we use IPPROTO_TCP.
 */
static int use_tcp;

/**
 * Protocol to use.
 */
static uint8_t proto;

/**
 * Test if all activities have finished, and if so,
 * terminate.
 */
static void
test_finished ()
{
  if (NULL != ah)
    return;
  if (NULL != nt)
    return;
  GNUNET_SCHEDULER_shutdown ();
}


/**
 * Function to iterate over sugested changes options
 *
 * @param cls closure
 * @param section name of the section
 * @param option name of the option
 * @param value value of the option
 */
static void
auto_conf_iter (void *cls,
                const char *section,
                const char *option,
                const char *value)
{
  struct GNUNET_CONFIGURATION_Handle *new_cfg = cls;

  printf ("%s: %s\n", option, value);
  if (NULL != new_cfg)
    GNUNET_CONFIGURATION_set_value_string (new_cfg, section, option, value);
}


/**
 * Function called with the result from the autoconfiguration.
 *
 * @param cls closure
 * @param diff minimal suggested changes to the original configuration
 *             to make it work (as best as we can)
 * @param result #GNUNET_NAT_ERROR_SUCCESS on success, otherwise the specific error code
 * @param type what the situation of the NAT
 */
static void
auto_config_cb (void *cls,
                const struct GNUNET_CONFIGURATION_Handle *diff,
                enum GNUNET_NAT_StatusCode result,
                enum GNUNET_NAT_Type type)
{
  const char *nat_type;
  char unknown_type[64];
  struct GNUNET_CONFIGURATION_Handle *new_cfg;

  ah = NULL;
  switch (type)
  {
  case GNUNET_NAT_TYPE_NO_NAT:
    nat_type = "NO NAT";
    break;

  case GNUNET_NAT_TYPE_UNREACHABLE_NAT:
    nat_type = "NAT but we can traverse";
    break;

  case GNUNET_NAT_TYPE_STUN_PUNCHED_NAT:
    nat_type = "NAT but STUN is able to identify the correct information";
    break;

  case GNUNET_NAT_TYPE_UPNP_NAT:
    nat_type = "NAT but UPNP opened the ports";
    break;

  default:
    sprintf (unknown_type, "NAT unknown, type %u", type);
    nat_type = unknown_type;
    break;
  }

  GNUNET_log (GNUNET_ERROR_TYPE_MESSAGE,
              "NAT status: %s/%s\n",
              GNUNET_NAT_AUTO_status2string (result),
              nat_type);

  if (NULL == diff)
    return;

  /* Shortcut: if there are no changes suggested, bail out early. */
  if (GNUNET_NO == GNUNET_CONFIGURATION_is_dirty (diff))
  {
    test_finished ();
    return;
  }

  /* Apply diff to original configuration and show changes
     to the user */
  new_cfg = write_cfg ? GNUNET_CONFIGURATION_dup (cfg) : NULL;

  GNUNET_log (GNUNET_ERROR_TYPE_MESSAGE,
              _ ("Suggested configuration changes:\n"));
  GNUNET_CONFIGURATION_iterate_section_values (diff,
                                               "nat",
                                               &auto_conf_iter,
                                               new_cfg);

  /* If desired, write configuration to file; we write only the
     changes to the defaults to keep things compact. */
  if (write_cfg)
  {
    struct GNUNET_CONFIGURATION_Handle *def_cfg;

    GNUNET_CONFIGURATION_set_value_string (new_cfg, "ARM", "CONFIG", NULL);
    def_cfg = GNUNET_CONFIGURATION_create ();
    GNUNET_break (GNUNET_OK == GNUNET_CONFIGURATION_load (def_cfg, NULL));
    if (GNUNET_OK !=
        GNUNET_CONFIGURATION_write_diffs (def_cfg, new_cfg, cfg_file))
    {
      GNUNET_log (GNUNET_ERROR_TYPE_MESSAGE,
                  _ ("Failed to write configuration to `%s'\n"),
                  cfg_file);
      global_ret = 1;
    }
    else
    {
      GNUNET_log (GNUNET_ERROR_TYPE_MESSAGE,
                  _ ("Wrote updated configuration to `%s'\n"),
                  cfg_file);
    }
    GNUNET_CONFIGURATION_destroy (def_cfg);
  }

  if (NULL != new_cfg)
    GNUNET_CONFIGURATION_destroy (new_cfg);
  test_finished ();
}


/**
 * Function called to report success or failure for
 * NAT configuration test.
 *
 * @param cls closure
 * @param result #GNUNET_NAT_ERROR_SUCCESS on success, otherwise the specific error code
 */
static void
test_report_cb (void *cls, enum GNUNET_NAT_StatusCode result)
{
  nt = NULL;
  printf ("NAT test result: %s\n", GNUNET_NAT_AUTO_status2string (result));
  test_finished ();
}


/**
 * Task run on shutdown.
 *
 * @param cls NULL
 */
static void
do_shutdown (void *cls)
{
  if (NULL != ah)
  {
    GNUNET_NAT_AUTO_autoconfig_cancel (ah);
    ah = NULL;
  }
  if (NULL != nt)
  {
    GNUNET_NAT_AUTO_test_stop (nt);
    nt = NULL;
  }
}


/**
 * Main function that will be run.
 *
 * @param cls closure
 * @param args remaining command-line arguments
 * @param cfgfile name of the configuration file used (for saving, can be NULL!)
 * @param c configuration
 */
static void
run (void *cls,
     char *const *args,
     const char *cfgfile,
     const struct GNUNET_CONFIGURATION_Handle *c)
{
  cfg_file = cfgfile;
  cfg = c;

  GNUNET_SCHEDULER_add_shutdown (&do_shutdown, NULL);

  if (do_auto)
  {
    ah = GNUNET_NAT_AUTO_autoconfig_start (c, &auto_config_cb, NULL);
  }

  if (use_tcp && use_udp)
  {
    if (do_auto)
      return;
    GNUNET_log (GNUNET_ERROR_TYPE_MESSAGE, "Cannot use TCP and UDP\n");
    global_ret = 1;
    return;
  }
  proto = 0;
  if (use_tcp)
    proto = IPPROTO_TCP;
  if (use_udp)
    proto = IPPROTO_UDP;

  if (NULL != section_name)
  {
    nt = GNUNET_NAT_AUTO_test_start (c,
                                     proto,
                                     section_name,
                                     &test_report_cb,
                                     NULL);
  }
  test_finished ();
}


/**
 * Main function of gnunet-nat-auto
 *
 * @param argc number of command-line arguments
 * @param argv command line
 * @return 0 on success, -1 on error
 */
int
main (int argc, char *const argv[])
{
  struct GNUNET_GETOPT_CommandLineOption options[] =
  { GNUNET_GETOPT_option_flag ('a',
                               "auto",
                               gettext_noop ("run autoconfiguration"),
                               &do_auto),

    GNUNET_GETOPT_option_string (
      'S',
      "section",
      "NAME",
      gettext_noop (
        "section name providing the configuration for the adapter"),
      &section_name),

    GNUNET_GETOPT_option_flag ('t', "tcp", gettext_noop ("use TCP"), &use_tcp),

    GNUNET_GETOPT_option_flag ('u', "udp", gettext_noop ("use UDP"), &use_udp),

    GNUNET_GETOPT_option_flag (
      'w',
      "write",
      gettext_noop ("write configuration file (for autoconfiguration)"),
      &write_cfg),
    GNUNET_GETOPT_OPTION_END };

  if (GNUNET_OK != GNUNET_STRINGS_get_utf8_args (argc, argv, &argc, &argv))
    return 2;
  if (GNUNET_OK !=
      GNUNET_PROGRAM_run (argc,
                          argv,
                          "gnunet-nat-auto [options]",
                          _ ("GNUnet NAT traversal autoconfiguration"),
                          options,
                          &run,
                          NULL))
  {
    global_ret = 1;
  }
  GNUNET_free ((void *) argv);
  return global_ret;
}


/* end of gnunet-nat-auto.c */
